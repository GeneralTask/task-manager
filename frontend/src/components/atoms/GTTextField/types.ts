export type FontSize = 'small' | 'medium' | 'large'

// props we support for markdown
export interface MarkdownEditorProps {
    value: string
    onChange: (newValue: string) => void
    placeholder?: string
    fontSize: FontSize
    type: 'markdown'
    itemId?: string // use if reusing field for multiple items
    autoFocus?: boolean
    autoSelect?: boolean
    blurOnEnter?: boolean
    disabled?: boolean
    isFullHeight?: boolean
    maxHeight?: number
    minHeight?: number
}

// all props we support for markdown + native textarea props
export interface PlainTextEditorProps
    extends Omit<MarkdownEditorProps, 'type'>,
        Omit<React.TextareaHTMLAttributes<HTMLTextAreaElement>, 'onChange' | 'value'> {
    type: 'plaintext'
}

export type GTTextFieldProps = MarkdownEditorProps | PlainTextEditorProps
