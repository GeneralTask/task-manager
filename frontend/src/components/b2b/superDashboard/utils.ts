import { Colors } from '../../../styles'
import { TLineColor } from './types'

export const getLineColor = (colorKey: TLineColor) => {
    switch (colorKey) {
        case 'pink':
            return Colors.control.primary.bg
        case 'gray':
            return Colors.background.hover
        case 'blue':
            return Colors.semantic.blue.base
    }
}
