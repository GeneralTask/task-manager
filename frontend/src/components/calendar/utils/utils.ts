import { DEFAULT_CALENDAR_COLOR, calendarColors } from './colors'

// backend sends empty string for title if it is the primary calendar, so fall back to account id
export const getCalendarName = (accountId: string, calendarTitle?: string): string => calendarTitle || accountId

export const getCalendarColor = (colorId: string): string =>
    calendarColors[colorId as keyof typeof calendarColors]?.background ?? DEFAULT_CALENDAR_COLOR
