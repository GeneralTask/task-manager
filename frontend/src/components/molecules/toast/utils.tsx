import { ToastType, toast } from 'react-hot-toast'
import { v4 as uuidv4 } from 'uuid'
import { Colors } from '../../../styles'
import { icons } from '../../../styles/images'
import Flex from '../../atoms/Flex'
import { TIconType } from '../../atoms/Icon'
import GTButton, { GTButtonProps } from '../../atoms/buttons/GTButton'

export interface EmitProps {
    toastId?: string
    title?: string
    message: string
    type?: ToastType
    duration?: number
    action?: GTButtonProps
    actions?: GTButtonProps[]
    undoAction?: {
        onClick: () => void
        onDismiss: () => void
    }
}
export const emit = (props: EmitProps) => {
    const { toastId, title, message, type, duration, action, actions, undoAction } = props

    const id = toastId ?? uuidv4()

    const toastContent = (
        <>
            <Flex column>
                {title && <strong>{title} </strong>}
                {message}
            </Flex>
            {action && (
                <div>
                    <GTButton {...action} />
                </div>
            )}
            {actions?.map((a, i) => (
                <div key={i}>
                    <GTButton {...a} />
                </div>
            ))}
            {undoAction && (
                <div>
                    <GTButton styleType="secondary" value="Undo" onClick={undoAction.onClick} />
                </div>
            )}
            {type !== 'loading' && (
                <div>
                    <GTButton
                        styleType="icon"
                        icon={icons.x}
                        onClick={() => {
                            undoAction?.onDismiss?.()
                            toast.dismiss(id)
                        }}
                    />
                </div>
            )}
        </>
    )

    const passProps = { id, duration }
    switch (type) {
        case 'success':
            return toast.success(toastContent, passProps)
        case 'error':
            return toast.error(toastContent, passProps)
    }
    return toast(toastContent, passProps)
}

export const getToastIcon = (type: ToastType): { icon: TIconType | null; iconColor: string | null } => {
    switch (type) {
        case 'blank':
            return { icon: icons.infoCircleSolid, iconColor: Colors.text.muted }
        case 'success':
            return { icon: icons.checkCircleSolid, iconColor: Colors.semantic.success.base }
        case 'error':
            return { icon: icons.exclamationTriangleSolid, iconColor: Colors.semantic.warning.base }
    }
    return { icon: null, iconColor: null }
}

export const getToastBgColor = (type: ToastType): string => {
    switch (type) {
        case 'blank':
            return Colors.background.base
        case 'success':
            return Colors.semantic.success.faint
        case 'error':
            return Colors.semantic.warning.faint
    }
    return Colors.background.border
}
