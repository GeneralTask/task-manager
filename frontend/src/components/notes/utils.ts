import { REACT_APP_NOTES_BASE_URL } from '../../constants'

export const getNoteURL = (id: string) => `${REACT_APP_NOTES_BASE_URL}/note/${id}`
