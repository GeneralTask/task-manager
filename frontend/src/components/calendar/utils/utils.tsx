import { TCalendarAccount } from '../../../utils/types'
import GTButton from '../../atoms/buttons/GTButton'
import { DEFAULT_CALENDAR_COLOR, calendarColors } from './colors'

// backend sends empty string for title if it is the primary calendar, so fall back to account id
export const getCalendarName = (accountId: string, calendarTitle?: string): string => calendarTitle || accountId

export const getCalendarColor = (colorId: string): string =>
    calendarColors[colorId as keyof typeof calendarColors]?.background ?? DEFAULT_CALENDAR_COLOR

export const getCalendarAuthButton = (
    account: TCalendarAccount,
    onClick: () => void,
    shortenLabels?: boolean,
    isAuthorizing?: boolean
) => {
    const getValue = () => {
        if (!account.has_primary_calendar_scopes && !account.has_multical_scopes) {
            if (isAuthorizing) {
                return 'Re-linking...'
            } else if (shortenLabels) {
                return 'Re-link'
            } else {
                return 'Re-link account'
            }
        }
        if (account.has_primary_calendar_scopes && !account.has_multical_scopes) {
            if (isAuthorizing) {
                return 'Authorizing...'
            } else if (shortenLabels) {
                return 'Authorize'
            } else {
                return 'Authorize all calendars'
            }
        }
    }
    if (!account.has_primary_calendar_scopes && !account.has_multical_scopes) {
        return (
            <GTButton
                onClick={onClick}
                value={getValue()}
                styleType="secondary"
                size="small"
                textColor="red"
                disabled={isAuthorizing}
            />
        )
    }
    if (account.has_primary_calendar_scopes && !account.has_multical_scopes) {
        return <GTButton value={getValue()} size="small" onClick={onClick} disabled={isAuthorizing} />
    }
    return null
}
