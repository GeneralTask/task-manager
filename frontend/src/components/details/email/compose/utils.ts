import { TEmail, TRecipients } from "../../../../utils/types"

import { EmailComposeType } from "../../../../utils/enums"
import { DateTime } from "luxon"

const prefixes = ['Re: ', 'Fwd: ', 'Fw: ']

export function stripSubjectPrefix(subject: string): string {
    for (const prefix of prefixes) {
        if (subject.startsWith(prefix)) {
            return subject.substring(prefix.length)
        }
    }
    return subject
}

export function attachSubjectPrefix(subject: string, composeType: EmailComposeType): string {
    let prefix = ''
    if (composeType === EmailComposeType.REPLY || composeType === EmailComposeType.REPLY_ALL) {
        prefix = 'Re: '
    }
    else if (composeType === EmailComposeType.FORWARD) {
        prefix = 'Fwd: '
    }
    return prefix + subject
}

export function getInitialRecipients(email: TEmail, composeType: EmailComposeType, userEmail: string): TRecipients {
    if (email.sender.email === userEmail) {
        if (composeType === EmailComposeType.REPLY) {
            return {
                to: email.recipients.to,
                cc: [],
                bcc: []
            }
        }
        else if (composeType === EmailComposeType.REPLY_ALL) {
            return {
                to: email.recipients.to,
                cc: email.recipients.cc.filter(r => r.email !== userEmail),
                bcc: []
            }
        }
    }
    else {
        if (composeType === EmailComposeType.REPLY) {
            return {
                to: [{ email: email.sender.email, name: '' }],
                cc: [],
                bcc: [],
            }
        }
        else if (composeType === EmailComposeType.REPLY_ALL) {
            return {
                to: [...new Set([
                    email.sender.email,
                    ...email.recipients.to.map(recipient => recipient.email).filter(email => email !== userEmail),
                    ...email.recipients.cc.map(recipient => recipient.email).filter(email => email !== userEmail),
                ])].map(email => ({ email, name: '' })),
                cc: [],
                bcc: [],
            }
        }
    }
    return {
        to: [],
        cc: [],
        bcc: [],
    }
}

export const emailsToRecipients = (emails: string[]) => emails.map(email => ({ email, name: '' }))

export const formatSentAtDateTime = (date: string) => DateTime.fromISO(date).toLocaleString(DateTime.DATETIME_MED)
