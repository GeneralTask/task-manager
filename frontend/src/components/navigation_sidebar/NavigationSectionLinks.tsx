import { useCallback, useEffect, useRef, useState } from 'react'
import { useParams } from 'react-router-dom'
import styled from 'styled-components'
import { v4 as uuidv4 } from 'uuid'
import { DEFAULT_SECTION_ID } from '../../constants'
import { useAddTaskSection, useModifyTaskSection } from '../../services/api/task-section.hooks'
import { useGetTasks } from '../../services/api/tasks.hooks'
import { Colors, Spacing, Typography } from '../../styles'
import { icons } from '../../styles/images'
import { DropItem, DropType, TTaskSection } from '../../utils/types'
import { Icon } from '../atoms/Icon'
import Loading from '../atoms/Loading'
import NoStyleInput from '../atoms/NoStyleInput'
import ReorderDropContainer from '../atoms/ReorderDropContainer'
import GTIconButton from '../atoms/buttons/GTIconButton'
import NavigationContextMenuWrapper from '../radix/NavigationContextMenuWrapper'
import IntegrationLinks from './IntegrationLinks'
import NavigationHeader from './NavigationHeader'
import NavigationLink, { NavigationLinkTemplate } from './NavigationLink'

const AddSectionContainer = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
    padding: ${Spacing._8};
    gap: ${Spacing._12};
    width: 100%;
    box-sizing: border-box;
`
const InputContainer = styled.div`
    & input {
        color: ${Colors.text.black};
        border: none;
        font-family: inherit;
        box-sizing: border-box;
        width: 100%;
        ${Typography.bodySmall};
    }
`

const NavigationSectionLinks = () => {
    const [isAddSectionInputVisible, setIsAddSectionInputVisible] = useState(false)
    const [sectionName, setSectionName] = useState('')
    const { mutate: addTaskSection } = useAddTaskSection()
    const { mutate: modifyTaskSection } = useModifyTaskSection()

    const { data: folders } = useGetTasks()
    const { section: sectionId } = useParams()

    const onKeyChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
        setSectionName(e.target.value)
    }
    const onKeyDownHandler = (e: React.KeyboardEvent<HTMLInputElement>) => {
        e.stopPropagation()
        if (e.key === 'Enter' && sectionName.trim() !== '') {
            setSectionName('')
            addTaskSection({ optimisticId: uuidv4(), name: sectionName, id_ordering: folders?.length })
            setIsAddSectionInputVisible(false)
        } else if (e.key === 'Escape' && inputRef.current) {
            setSectionName('')
            setIsAddSectionInputVisible(false)
        }
    }
    const onBlurHandler = () => {
        if (sectionName.trim() !== '') {
            addTaskSection({ optimisticId: uuidv4(), name: sectionName, id_ordering: folders?.length })
        }
        setIsAddSectionInputVisible(false)
        setSectionName('')
    }

    const inputRef = useRef<HTMLInputElement>(null)
    const onOpenAddSectionInputHandler = useCallback(
        (e: React.MouseEvent<HTMLButtonElement>) => {
            e.stopPropagation()
            setIsAddSectionInputVisible(true)
            inputRef.current?.focus()
        },
        [inputRef]
    )

    useEffect(() => {
        if (isAddSectionInputVisible && inputRef.current) {
            inputRef.current.focus()
        }
    }, [isAddSectionInputVisible])

    const handleClickOutsideInput = (e: MouseEvent) => {
        if (!inputRef.current?.contains(e.target as Node)) {
            setIsAddSectionInputVisible(false)
        }
    }
    useEffect(() => {
        document.addEventListener('click', handleClickOutsideInput, false)
        return () => {
            document.removeEventListener('click', handleClickOutsideInput, false)
        }
    }, [])

    const handleReorder = useCallback((item: DropItem, dropIndex: number) => {
        modifyTaskSection(
            {
                id: item.id,
                id_ordering: dropIndex,
            },
            item.task?.optimisticId
        )
    }, [])

    const defaultFolder = folders?.find((section) => section.id === DEFAULT_SECTION_ID)
    const doneFolder = folders?.find((section) => section.is_done)
    const trashFolder = folders?.find((section) => section.is_trash)

    // Logic for updating section name from navigation view
    const [sectionBeingEdited, setSectionBeingEdited] = useState<TTaskSection | null>(null)
    const [updatedSectionName, setUpdatedSectionName] = useState<string>('')
    const ref = useRef<HTMLDivElement>(null)

    const onKeyDownHandlerSectionName = (e: React.KeyboardEvent<HTMLInputElement>) => {
        e.stopPropagation()
        if (sectionBeingEdited == null) return
        if (e.key === 'Enter' && updatedSectionName.trim() !== '') {
            setSectionBeingEdited(null)
            setUpdatedSectionName('')
            modifyTaskSection(
                {
                    id: sectionBeingEdited.id,
                    name: updatedSectionName,
                },
                sectionBeingEdited.optimisticId
            )
        } else if (e.key === 'Escape') {
            setUpdatedSectionName('')
            setSectionBeingEdited(null)
        }
    }

    if (!folders) {
        return <Loading />
    }

    return (
        <>
            <IntegrationLinks />
            <NavigationHeader
                title="Folders"
                rightContent={
                    <GTIconButton icon={icons.plus} onClick={onOpenAddSectionInputHandler} tooltipText="Add Folder" />
                }
            />
            {defaultFolder && (
                <NavigationLink
                    link={`/tasks/${defaultFolder.id}`}
                    title={defaultFolder.name}
                    icon={icons.inbox}
                    isCurrentPage={sectionId === defaultFolder.id}
                    taskSection={defaultFolder}
                    count={defaultFolder.tasks.length}
                    droppable
                />
            )}
            {folders
                ?.filter((section) => section.id !== DEFAULT_SECTION_ID && !section.is_done && !section.is_trash)
                .map((section, index) =>
                    sectionBeingEdited?.id !== section.id ? (
                        <ReorderDropContainer
                            key={section.id}
                            index={index} // +1 because we skip the default folder
                            acceptDropType={DropType.FOLDER}
                            onReorder={handleReorder}
                        >
                            <NavigationContextMenuWrapper
                                section={section}
                                setSectionBeingEdited={(section) => {
                                    setUpdatedSectionName(section.name)
                                    setSectionBeingEdited(section)
                                }}
                            >
                                <NavigationLink
                                    link={`/tasks/${section.id}`}
                                    title={section.name}
                                    icon={icons.folder}
                                    isCurrentPage={sectionId === section.id}
                                    taskSection={section}
                                    count={section.tasks.length}
                                    draggable
                                    droppable
                                />
                            </NavigationContextMenuWrapper>
                        </ReorderDropContainer>
                    ) : (
                        <NavigationLinkTemplate key={section.id} ref={ref}>
                            <AddSectionContainer>
                                <div>
                                    <Icon icon={icons.folder} color="black" />
                                </div>
                                <InputContainer>
                                    <NoStyleInput
                                        ref={(node) => {
                                            if (!node) return
                                            if (sectionBeingEdited.id === section.id) node.focus()
                                            else if (sectionBeingEdited === null) node.blur()
                                        }}
                                        value={updatedSectionName}
                                        onChange={(e) => {
                                            setUpdatedSectionName(e.target.value)
                                        }}
                                        onKeyDown={onKeyDownHandlerSectionName}
                                        placeholder="Enter a section name"
                                    />
                                </InputContainer>
                            </AddSectionContainer>
                        </NavigationLinkTemplate>
                    )
                )}
            {isAddSectionInputVisible && (
                <NavigationLinkTemplate>
                    <AddSectionContainer>
                        <div>
                            <Icon icon={icons.folder} color="black" />
                        </div>
                        <InputContainer>
                            <NoStyleInput
                                ref={inputRef}
                                value={sectionName}
                                onChange={onKeyChangeHandler}
                                onKeyDown={onKeyDownHandler}
                                placeholder="Add Folder"
                                onBlur={onBlurHandler}
                            />
                        </InputContainer>
                    </AddSectionContainer>
                </NavigationLinkTemplate>
            )}
            <ReorderDropContainer
                index={folders.length - 2} // -2 because we skip the done and trash folders
                acceptDropType={DropType.FOLDER}
                onReorder={handleReorder}
                indicatorType="TOP_ONLY"
            >
                <>
                    {doneFolder && (
                        <NavigationLink
                            link={`/tasks/${doneFolder.id}`}
                            title={doneFolder.name}
                            icon={icons.checkbox_checked}
                            isCurrentPage={sectionId === doneFolder.id}
                            count={doneFolder.tasks.length}
                            taskSection={doneFolder}
                            droppable
                        />
                    )}
                    {trashFolder && (
                        <NavigationLink
                            link={`/tasks/${trashFolder.id}`}
                            title={trashFolder.name}
                            icon={icons.trash}
                            isCurrentPage={sectionId === trashFolder.id}
                            taskSection={trashFolder}
                            droppable
                        />
                    )}
                </>
            </ReorderDropContainer>
        </>
    )
}

export default NavigationSectionLinks
