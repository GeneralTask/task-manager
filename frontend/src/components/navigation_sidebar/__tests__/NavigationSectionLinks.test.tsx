/**
 * @jest-environment jsdom
 */

import React from 'react'
import renderer from 'react-test-renderer'
import { TTaskSection } from '../../../utils/types'
import NavigationSectionLinks from '../NavigationSectionLinks'
import { QueryClient, QueryClientProvider } from 'react-query'
import { DndProvider } from 'react-dnd'
import { HTML5Backend } from 'react-dnd-html5-backend'
import { BrowserRouter } from 'react-router-dom'


const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            refetchOnWindowFocus: false,
        },
    },
})

const noTaskSections: TTaskSection[] = []
const singleTaskSection: TTaskSection[] = [{
    id: '1',
    name: 'Section 1',
    tasks: [],
    is_done: false,
}]

test('NavigationSectionLinks renders three components when there are 0 task sections', () => {
    const tree = renderer.create(
        <QueryClientProvider client={queryClient}>
            <DndProvider backend={HTML5Backend}>
                <BrowserRouter>
                    <NavigationSectionLinks taskSections={noTaskSections} sectionId={''} pathName={''} />
                </BrowserRouter>
            </DndProvider>
        </QueryClientProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
})


test('NavigationSectionLinks renders three components when there is 1 task section', () => {
    const tree = renderer.create(
        <QueryClientProvider client={queryClient}>
            <DndProvider backend={HTML5Backend}>
                <BrowserRouter>
                    <NavigationSectionLinks taskSections={singleTaskSection} sectionId={''} pathName={''} />
                </BrowserRouter>
            </DndProvider>
        </QueryClientProvider>
    ).toJSON()
    expect(tree).toMatchSnapshot()
})
