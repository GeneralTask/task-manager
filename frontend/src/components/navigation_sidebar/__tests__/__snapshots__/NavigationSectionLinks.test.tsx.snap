// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationSectionLinks renders three components when there are 0 task sections 1`] = `
Array [
  <View />,
  <View />,
]
`;

exports[`NavigationSectionLinks renders three components when there is 1 task section 1`] = `
Array [
  <View />,
  <View />,
  <View />,
]
`;
